Sources/PackageRegistry/RegistryManager.swift

```diff
@@ -0,0 +1,253 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import PackageLoading
+import PackageModel
+
+import TSCBasic
+import TSCUtility
+
+import Foundation
+#if canImport(FoundationNetworking)
+import FoundationNetworking
+#endif
+
+public enum RegistryError: Error {
+    case invalidOperation
+    case invalidResponse
+    case invalidURL
+    case invalidChecksum(expected: String, actual: String)
+    case invalidPackage
+}
+
+public final class RegistryManager {
+    public static var configuration: URLSessionConfiguration = {
+        var configuration = URLSessionConfiguration.default
+        configuration.httpAdditionalHeaders = [
+            "User-Agent": "Swift Package Manager",
+        ]
+
+        return configuration
+    }()
+
+    private lazy var session = URLSession(configuration: Self.configuration)
+
+    let registryBaseURL: Foundation.URL
+
+    init(registryBaseURL: Foundation.URL) {
```

Comment:
just an idea, but if the call site wants to enforce specific dispatch queue for the completion handlers to run on we can take one in the ctor or in the individual functions